Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm

```diff
@@ -809,6 +803,201 @@ static Color checkboxRadioBorderColorForVectorBasedControls(OptionSet<ControlSty
     return backgroundColor;
 }
 
+static RoundedShape continuousRoundedShape(const FloatRect& rect, const float cornerRadius, ShouldComputePath computePath)
+{
+    if (computePath == ShouldComputePath::No)
+        return { { }, rect, cornerRadius, CornerType::Continuous };
+
+    Path path;
+    path.addContinuousRoundedRect(rect, cornerRadius);
+    return { path, rect, cornerRadius, CornerType::Continuous };
+}
+
+static RoundedShape roundedShape(const FloatRect& rect, const float cornerRadius, ShouldComputePath computePath)
+{
+    if (computePath == ShouldComputePath::No)
+        return { { }, rect, cornerRadius, CornerType::Noncontinuous };
+
+    Path path;
+    path.addRoundedRect(FloatRoundedRect { rect, FloatRoundedRect::Radii { cornerRadius, cornerRadius } });
+    return { path, rect, cornerRadius, CornerType::Noncontinuous };
+}
+
+static bool nodeIsDateOrTimeRelatedInput(Node* node)
+{
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(node)) {
+        return input->isDateField()
+        || input->isDateTimeLocalField()
+        || input->isMonthField()
+        || input->isTimeField()
+        || input->isWeekField();
+    }
+
+    return false;
+}
+
+static RoundedShape shapeForButton(const RenderObject& box, const FloatRect& rect, ShouldComputePath computePath = ShouldComputePath::Yes)
+{
+    CheckedRef style = box.style();
+    const auto zoomScale = style->usedZoom();
+
+    auto controlRadius = 0.f;
+#if PLATFORM(MAC)
+    controlRadius = 6.f * zoomScale;
+#else
+    static constexpr auto largeButtonHeight = 45;
+    static constexpr auto largeButtonRadiusRatio = 0.35f / 2;
+
+    const auto isVertical = !style->writingMode().isHorizontal();
+    const auto boxLogicalHeight = isVertical ? rect.width() : rect.height();
+    const auto minDimension = std::min(rect.width(), rect.height());
+
+    const auto radiusForNonDateRelatedLargeButton = minDimension * largeButtonRadiusRatio;
+    if (nodeIsDateOrTimeRelatedInput(box.node()))
+        controlRadius = 5.f * zoomScale;
+    else if (boxLogicalHeight >= largeButtonHeight * zoomScale)
+        controlRadius = radiusForNonDateRelatedLargeButton;
+    else {
+        controlRadius = minDimension / 2.f;
+
+        // If trying to make the button pill-shaped would make it a circle
+        // or nearly circle, use the non-pill shape instead.
+        const auto sizeRatio = rect.width() / rect.height();
+        if (3 / 2.f > sizeRatio && sizeRatio > 2 / 3.f)
```

Comment:
Might be worth pulling out 1.5 and 0.67 into named constants? I'm not sure what they mean tbh, maybe it's not that important though.

Reply:
In that case, maybe 1.5 could be the constant, and you could use 1/x in the other?